// ngx-loading/esm2015/lib/ngx-loading-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function INgxLoadingConfig() { }
if (false) {
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.backdropBorderRadius;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.backdropBackgroundColour;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.fullScreenBackdrop;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.animationType;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.primaryColour;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.secondaryColour;
    /** @type {?|undefined} */
    INgxLoadingConfig.prototype.tertiaryColour;
    /* Skipping unhandled member: [key: string]: string | boolean | undefined;*/
}
export class NgxLoadingConfig {
    /**
     * @param {?=} config
     */
    constructor(config = {}) {
        this.backdropBorderRadius = config.backdropBorderRadius;
        this.backdropBackgroundColour = config.backdropBackgroundColour;
        this.fullScreenBackdrop = config.fullScreenBackdrop;
        this.animationType = config.animationType;
        this.primaryColour = config.primaryColour;
        this.secondaryColour = config.secondaryColour;
        this.tertiaryColour = config.tertiaryColour;
    }
}
if (false) {
    /** @type {?} */
    NgxLoadingConfig.prototype.backdropBorderRadius;
    /** @type {?} */
    NgxLoadingConfig.prototype.backdropBackgroundColour;
    /** @type {?} */
    NgxLoadingConfig.prototype.fullScreenBackdrop;
    /** @type {?} */
    NgxLoadingConfig.prototype.animationType;
    /** @type {?} */
    NgxLoadingConfig.prototype.primaryColour;
    /** @type {?} */
    NgxLoadingConfig.prototype.secondaryColour;
    /** @type {?} */
    NgxLoadingConfig.prototype.tertiaryColour;
    /* Skipping unhandled member: [key: string]: string | boolean | undefined;*/
}
/** @type {?} */
export const ngxLoadingAnimationTypes = {
    chasingDots: 'chasing-dots',
    circle: 'sk-circle',
    circleSwish: 'circleSwish',
    cubeGrid: 'sk-cube-grid',
    doubleBounce: 'double-bounce',
    none: 'none',
    pulse: 'pulse',
    rectangleBounce: 'rectangle-bounce',
    rotatingPlane: 'rotating-plane',
    threeBounce: 'three-bounce',
    wanderingCubes: 'wandering-cubes'
};
